+nosnoc/+solver/MpccSolver.m

```diff
@@ -72,16 +69,294 @@
+                    if opts.decreasing_s_elastic_upper_bound
+                        nlp.g.s_ub = {nlp.w.s_elastic() - nlp.p.sigma_p(), -inf, 0};
+                    end
+                    if opts.objective_scaling_direct
+                        nlp.f = nlp.f + (1/nlp.p.sigma_p())*nlp.w.s_elastic(); % penalize the elastic more and more with decreasing sigma_p
+                    else
+                        nlp.f = nlp.p.sigma_p()*nlp.f + nlp.w.s_elastic(); % reduce the weight of the initial objective with decreasing sigma_p
+                    end
+                  case HomotopySteeringStrategy.ELL_1
+                    % Ell 1 steering strategy slacks are added during creation of constraints
+                end
+
+
+                psi_fun = get_psi_fun(MpccMethod(obj.relaxation_type), opts.normalize_homotopy_update);
+                comp_var_names = mpcc.G.get_var_names();
+                sum_elastic = 0;
+                for ii=1:numel(comp_var_names)
+                    name = comp_var_names{ii};
+                    depth = mpcc.G.(name).depth;
+                    s_elastic_name = ['s_elastic_' char(name)];
+                    if depth == 0                        
+                        G_curr = mpcc.G.(name)();
+                        H_curr = mpcc.H.(name)();
+                        % select relaxation slacks/parameters
+                        switch opts.homotopy_steering_strategy
+                          case HomotopySteeringStrategy.DIRECT
+                            % nlp.p.sigma_p(): sigma is a parameter/variable that has no indices
+                            sigma = nlp.p.sigma_p(); 
+                          case HomotopySteeringStrategy.ELL_INF
+                            sigma = nlp.w.s_elastic(); % Here s_elastic takes the role of sigma in direct, and sigma_p is used to define a penalty parameter for the elastic variable s_elastic
+                          case HomotopySteeringStrategy.ELL_1
+                            % adding elastic variables to nlp.w which augments the original mpcc.w
+                            % Remark: ELL_1 with s_elastic is equivalent to usually Ell_1 penality approach, but this indirect way helps
+                            % to add some constraint on s_elastic (which  avoids unbounded problems somtimes, and it can also improve convergence)
+                            nlp.w.(s_elastic_name) = {{s_elastic_name, size(G_curr, 1)}, opts.s_elastic_min, opts.s_elastic_max, opts.s_elastic_0};
+                            if opts.decreasing_s_elastic_upper_bound
+                                % TODO(@anton) this creates a performance bottleneck
+                                nlp.g.([s_elastic_name '_ub']) = {nlp.w.(s_elastic_name)() - nlp.p.sigma_p(), -inf, 0};
+                            end
+
+                            % TODO(@anton) this creates a performance bottleneck
+                            sigma = nlp.w.(s_elastic_name)();
```

Comment:
an alternative would be in ell1, to not lift, and to not have slacks, as already all comps are linear/affine and they can be directly added to the objective.

Reply:
This is true for `nosnoc` but is not enforced generally. We can add this as an option in the future.